modules/gdscript/gdscript_function.cpp

```diff
@@ -1552,19 +1552,16 @@ Variant GDScriptFunctionState::_signal_callback(const Variant **p_args, int p_ar
 		GDScriptFunctionState *gdfs = Object::cast_to<GDScriptFunctionState>(ret);
 		if (gdfs && gdfs->function == function) {
 			completed = false;
-			gdfs->previous_state = Ref<GDScriptFunctionState>(this);
+			gdfs->first_state = first_state.is_valid() ? first_state : Ref<GDScriptFunctionState>(this);
 		}
 	}
 
 	function = NULL; //cleaned up;
 	state.result = Variant();
 
 	if (completed) {
-		GDScriptFunctionState *state = this;
-		while (state != NULL) {
-			state->emit_signal("completed", ret);
-			state = *(state->previous_state);
-		}
+		if (first_state.is_valid()) first_state->emit_signal("completed", ret);
```

Comment:
Please don't put condition and code block inline for `if`s, it makes it hard to read what is the condition and what is the statement that will be executed. We tolerate it for simple `if (cond) return false;`, but even then I'd advise not to use it.

Reply:
fixed